napalm_logs/server.py

```diff
@@ -28,36 +29,78 @@ def __init__(self,
         self.__pipe = pipe
         self.__os_pipe_map = os_pipe_map
         self.__up = False
+        self.compiled_prefixes = None
+
+    def _compile_prefixes(self):
+        '''
+        Create a dict of all OS prefixes and their compiled regexs
+        '''
+        self.compiled_prefixes = {}
+        for os, os_config in self.config.iteritems():
+            if not os_config:
+                continue
+            values = os_config.get('prefix', {}).get('values', {})
+            line = os_config.get('prefix', {}).get('line', '')
+            # We will now figure out which position each value is in so we can use it with the match statement
+            position = {}
+            for key in values.keys():
+                position[line.find('{' + key + '}')] = key
+            sorted_position = {}
+            for i, elem in enumerate(sorted(position.items())):
+                sorted_position[elem[1]] = i + 1
+
+            escaped = re.escape(line).replace('\{', '{').replace('\}', '}')
+            self.compiled_prefixes[os] = { 
+                'prefix': re.compile(escaped.format(**values)),
+                'prefix_positions': sorted_position,
+                'values': values,
+                'messages': {}
+                }
 
     def _identify_os(self, msg):
         '''
         Using the prefix of the syslog message,
         we are able to identify the operating system and then continue parsing.
         '''
-        return ('junos', 'crap')
+        ret = {}
+        for os, data in self.compiled_prefixes.iteritems():
+            match = data.get('prefix', '').search(msg)
+            positions = data.get('prefix_positions', {}) 
+            mapping = data.get('mapping')
+            values = data.get('values')
+            if not match:
+                continue
+            ret = {}
+            for key in values.keys():
+                ret[key] = match.group(positions.get(key))
+            # TODO Should we stop searching and just return, or should we return all matches OS?
+            return os, ret
+
+        log.debug('No OS matched for: %s', msg)
```

Comment:
Should return empty obj when fails finding the OS type.
And please use `format` instead of `%s` :D